Fix thread safety in getCurrentCUDABlasHandle and getCUDABlasLtWorkspace

t-ivan-gr opened the pull request:
Summary:
getCurrentCUDABlasHandle() and getCUDABlasLtWorkspace() use static mutable maps that are not protected from concurrent read-and-write. This leads to crashes.

This diff adds mutexes to synchronize access to the static maps.

Test Plan:
Use a GPU OD, run multi-threaded tests with TSAN:
```
buck test fbcode//mode/dev-tsan fbcode//caffe2:cuda_cublas_handle_pool_test  -- --stress-runs 100
```
https://www.internalfb.com/intern/testinfra/testrun/14355223937501118


TSAN: P2026731804

Differential Revision: D86316117

Skylion007 commented on aten/src/ATen/cuda/CublasHandlePool.cpp:
```diff
@@ -241,20 +249,45 @@ void* getCUDABlasLtWorkspace() {
+    auto workspace_it = workspace.map.find(key);
+    if (workspace_it == workspace.map.end()) {
+      workspace_it =
+          workspace.map.insert(workspace_it, {key, std::move(new_workspace)});
```
Not that it matters much, but you could probably use emplace here using std::piecewise_construct

t-ivan-gr replied:
I migrated to using .emplace() without the first hint paramter, since the hint will be pointing to map.end() anyway. This means we don't need to use piecewise_construct, simplifying things a bit.

Skylion007 commented on aten/src/ATen/cuda/CublasHandlePool.cpp:
```diff
@@ -300,11 +333,39 @@ cublasHandle_t getCurrentCUDABlasHandle() {
+    auto workspace_it = workspace.map.find(key);
+    if (workspace_it == workspace.map.end()) {
+      workspace_it =
+          workspace.map.insert(workspace_it, {key, std::move(new_workspace)});
```
Same here

ngimel commented on aten/src/ATen/cuda/CublasHandlePool.cpp:
```diff
@@ -241,20 +249,45 @@ void* getCUDABlasLtWorkspace() {
+
+  // Fast path: check if workspace already exists
+  {
+    std::lock_guard<std::mutex> lock(workspace.mutex);
```
this acquires full mutex in the hot path, you should use shared_mutex to make read-only accesses faster

ngimel commented:
btw @eqy why don't we add per-stream workspaces as attribute to the handle itself? Then, since handle is guaranteed to be unique per live thread, workspace lookout given a handle won't need any locks at all.

cyyever commented:
Can we switch to use `thread_local` handles and avoid all added mutexes?

eqy commented:
@ngimel I think that's a reasonable suggestion though not sure I want to hijack this PR

The only other consideration I'm thinking about is I'm not sure if we can also migrate ROCm logic to the same thing as IIRC their unified workspace path was different (non-Lt handle could not be used by Lt to get the workspace?)? https://github.com/pytorch/pytorch/blob/59307ca1bc256ad03140c3c55698763c75487c63/aten/src/ATen/cuda/CublasHandlePool.cpp#L236

malfet commented on aten/src/ATen/test/cuda_cublas_handle_pool_test.cpp:
```diff
@@ -0,0 +1,155 @@
+
+        for (int j = 0; j < iterations_per_thread; ++j) {
+          // This should not crash or cause data races
+          cublasHandle_t handle = at::cuda::getCurrentCUDABlasHandle();
```
```suggestion
          const auto handle = at::cuda::getCurrentCUDABlasHandle();
```

t-ivan-gr committed:
Fix thread safety in getCurrentCUDABlasHandle and getCUDABlasLtWorkspace (#167248)

Summary:

getCurrentCUDABlasHandle() and getCUDABlasLtWorkspace() use static mutable maps that are not protected from concurrent read-and-write. This leads to crashes.

This diff adds mutexes to synchronize access to the static maps.

Test Plan:
Use a GPU OD, run multi-threaded tests (cuda_cublas_handle_pool_test) with TSAN:
```
buck test fbcode//mode/dev-tsan fbcode//caffe2:cuda_cublas_handle_pool_test  -- --stress-runs 100
```
https://www.internalfb.com/intern/testinfra/testrun/14355223937501118


TSAN output (before synchronization was added): P2026731804

Reviewed By: malfet

Differential Revision: D86316117

t-ivan-gr commented:
@pytorchbot merge -i

jeffdaily commented:
Closing, since it was relanded by https://github.com/pytorch/pytorch/pull/167928.
